api/src/main/java/io/smallrye/stork/api/Service.java

```diff
@@ -12,39 +15,110 @@
  */
 public class Service {
 
+    private final Lock instanceSelectionLock;
     private final Optional<LoadBalancer> loadBalancer;
     private final ServiceDiscovery serviceDiscovery;
     private final String serviceName;
     private final boolean secure;
 
-    public Service(String serviceName, Optional<LoadBalancer> loadBalancer, ServiceDiscovery serviceDiscovery, boolean secure) {
+    public Service(String serviceName, Optional<LoadBalancer> loadBalancer, ServiceDiscovery serviceDiscovery,
+            boolean secure, boolean threadSafeLoadBalancer) {
         this.loadBalancer = loadBalancer;
         this.serviceDiscovery = serviceDiscovery;
         this.serviceName = serviceName;
         this.secure = secure;
+        instanceSelectionLock = threadSafeLoadBalancer ? null : new ReentrantLock();
     }
 
     /**
      * Selects a service instance.
+     * <p>
+     * The selection looks for the service instances and select the one to use using the load balancer.
+     *
+     * <b>Note:</b> this method doesn't record a start of an operation using this load balancer and does not
+     * synchronize load balancer invocations even if the load balancer is not thread safe
      *
+     * @return a Uni with a ServiceInstance, or with {@link NoServiceInstanceFoundException} if the load balancer failed to find
+     *         a service instance capable of handling a call
+     * @throws IllegalArgumentException if the current service does not use a load balancer
```

Comment:
Can this happen? It should use the round-robin LB by default.

Reply:
true, I forgot about it

Reply:
I'll change Optional<LoadBalancer> to LoadBalancer for the LB field type here too